geoopt/manifolds/birkhoff_polytope.py

```diff
@@ -0,0 +1,167 @@
+        col_sum = x.sum(dim=-2)
+        row_ok = torch.allclose(row_sum, row_sum.new((1,)).fill_(1), atol=atol, rtol=rtol)
+        col_ok = torch.allclose(col_sum, col_sum.new((1,)).fill_(1), atol=atol, rtol=rtol)
+        if row_ok and col_ok:
+            return True, None
+        else:
+            return False, "illegal doubly stochastic matrix with atol={}, rtol={}".format(atol, rtol)
+
+    def _check_vector_on_tangent(self, x, u, *, atol=1e-5, rtol=1e-5):
+        diff = u.transpose(-1, -2) @ x + x.transpose(-1, -2) @ u
+        ok = torch.allclose(diff, diff.new((1,)).fill_(0), atol=atol, rtol=rtol)
+        if not ok:
+            return False, "`u^T x + x^T u !=0` with atol={}, rtol={}".format(atol, rtol)
+        return True, None
+
+    def projx(self, x):
+        iter = 0
+        x_shape = x.shape
+        x_reshaped = x.reshape(-1, x_shape[-2], x_shape[-1])
+        c = 1.0 / (torch.sum(x_reshaped, dim=-2, keepdim=True) + self.epsilon)
+        r = 1.0 / (torch.matmul(x_reshaped, torch.transpose(c, -1, -2)) + self.epsilon)
+        while iter < self.maxiter:
+            iter += 1
+            cinv = torch.matmul(r.permute(0, 2, 1), x_reshaped)
+            if torch.max(torch.abs(cinv * c - 1)) <= self.tol:
+                break
+            c = 1.0 / (cinv + self.epsilon)
+            r = 1.0 / (torch.matmul(x_reshaped, torch.Tensor.permute(c, 0, 2, 1)) + self.epsilon)
+        return (x_reshaped * torch.matmul(r, c)).reshape(x_shape)
+        # x = x.reshape(x_shape)
+        # return x
+
+    def proju(self, x, u):
+        # takes batch data
+        # batch_size, n, _ = x.shape
+        x_shape = x.shape
+        x = x.reshape(-1, x_shape[-2], x_shape[-1])
+        batch_size, n = x.shape[0:2]
+
+        e = torch.ones(batch_size, n, 1)
+        I = torch.unsqueeze(torch.eye(x.shape[-1]), 0).repeat(batch_size, 1, 1)
+
+        mu = x * u
+
+        A = linalg.block_matrix([[I, x], [torch.transpose(x, 1, 2), I]])
+
+        B = A[:, :, 1:]
+        b = torch.cat(
+            [
+                torch.sum(mu, dim=2, keepdim=True),
+                torch.transpose(torch.sum(mu, dim=1, keepdim=True), 1, 2)
+            ],
+            dim=1
+        )
+
+        zeta, _ = torch.solve(B.transpose(1, 2) @ (b - A[:, :, 0:1]), B.transpose(1, 2) @ B)
+        alpha = torch.cat([torch.ones(batch_size, 1, 1), zeta[:, 0:n - 1]], dim=1)
+        beta = zeta[:, n - 1: 2 * n - 1]
+        rgrad = mu - (alpha @ e.transpose(1, 2) + e @ beta.transpose(1, 2)) * x
+
+        rgrad = rgrad.reshape(x_shape)
+        return rgrad
+
+    egrad2rgrad = proju
+
+    def retr(self, x, u):
+        K = u / x
+        Y = x * torch.exp(K)
+        Y = self.projx(Y)
+        Y = torch.max(Y, Y.new(1).fill_(1e-12))
+        return Y
+
+    expmap = retr
+
+    def inner(self, x, u, v=None, *, keepdim=False):
+        if v is None:
+            v = u
+        batch_size = x.shape[0]
+        return torch.sum(u * v / x) / batch_size
```

Comment:
- `keepdim` is not used here
- maybe `batch_size` is little confusing here

Reply:
The return shape seems to be scalar-valued, while for multiple points we return inner product per point.

Reply:
still `batch_size` is little confusing, maybe `n = x.shape[0]` would be better (when I see `batch_size`, it is always something related to optimization/nn, much more reasonable to keep the logic separate)